File: api-clients/elixir/lib/otreeba_open_cannabis_api/api/edibles.ex
defmodule OtreebaOpenCannabisAPI.Api.Edibles do
  @moduledoc """
  Documentation for OtreebaOpenCannabisAPI.Api.Edibles.
  """

  use Tesla

  plug Tesla.Middleware.BaseUrl, "https://api.otreeba.com/v1"
  plug Tesla.Middleware.JSON

  @doc """
  Find edible by Open Cannabis Product Code (OCPC).

  Returns a single edible.
  """
  def get_edible_by_ocpc(ocpc) do
    method = [method: :get]
    url = [url: "/edibles/#{ocpc}"]
    query_params = []
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end

  @doc """
  Get a list of all current edibles.

  Returns a paginated list of edibles.
  """
  def get_edibles(page, count, sort) do
    method = [method: :get]
    url = [url: "/edibles"]
    query_params = [query: [{:"page", page}, {:"count", count}, {:"sort", sort}]]
    header_params = []
    body_params = []
    form_params = []
    params = query_params ++ header_params ++ body_params ++ form_params
    opts = []
    options = method ++ url ++ params ++ opts

    request(options)
  end
end

File: api-clients/clojure/src/otreeba_open_cannabis_api/api/edibles.clj
(ns otreeba-open-cannabis-api.api.edibles
  (:require [otreeba-open-cannabis-api.core :refer [call-api check-required-params with-collection-format]])
  (:import (java.io File)))

(defn get-edible-by-ocpc-with-http-info
  "Find edible by Open Cannabis Product Code (OCPC).
  Returns a single edible."
  [ocpc ]
  (call-api "/edibles/{ocpc}" :get
            {:path-params   {"ocpc" ocpc }
             :header-params {}
             :query-params  {}
             :form-params   {}
             :content-types ["application/json"]
             :accepts       ["application/json"]
             :auth-names    []}))

(defn get-edible-by-ocpc
  "Find edible by Open Cannabis Product Code (OCPC).
  Returns a single edible."
  [ocpc ]
  (:data (get-edible-by-ocpc-with-http-info ocpc)))

(defn get-edibles-with-http-info
  "Get a list of all current edibles.
  Returns a paginated list of edibles."
  ([] (get-edibles-with-http-info nil))
  ([{:keys [page count sort ]}]
   (call-api "/edibles" :get
             {:path-params   {}
              :header-params {}
              :query-params  {"page" page "count" count "sort" sort }
              :form-params   {}
              :content-types ["application/json"]
              :accepts       ["application/json"]
              :auth-names    []})))

(defn get-edibles
  "Get a list of all current edibles.
  Returns a paginated list of edibles."
  ([] (get-edibles nil))
  ([optional-params]
   (:data (get-edibles-with-http-info optional-params))))

